Add EIP-5344: Soulmates Token (#5344)
ethereum/EIPs: EIPS/eip-1313.md

---

> eip: 1313

Reviewer: ```suggestion
eip: 5344
```
Also rename the file.

Author: Thank you for lot of feed back, I will update it.

title: Soulmates Token

> description: Bind souls(魂), tokens representing mates.

Reviewer 2: Needs to be more descriptive

> author: Jun Kawasaki

Reviewer 2: ```suggestion
author: Jun Kawasaki (@jun784)
```

Author: Thank you for feedback @pandapip1-san, I will update and improve it.

> discussions-to: [https://ethereum-magicians.org/t/eip-5114-soulbound-token/9417](https://ethereum-magicians.org/t/eip-1313-soulmate-token/10132)

Reviewer: ```suggestion
discussions-to: https://ethereum-magicians.org/t/eip-1313-soulmate-token/10132
```

> discussions-to: [https://ethereum-magicians.org/t/eip-5114-soulbound-token/9417](https://ethereum-magicians.org/t/eip-1313-soulmate-token/10132)

Reviewer 2: ```suggestion
discussions-to: https://ethereum-magicians.org/t/eip-1313-soulmate-token/10132
```

status: Draft
type: Standards Track
category: ERC
created: 2022-07-27
requires: 721, 5114
---

## Abstract

> A soulmates token is a countable soulbound token. when it is minted, and cannot be transferred/moved after that. but its can be add new mate anytime.

Reviewer 2: ```suggestion
A soulmates token is a countable soulbound (non-transferrable) non-fungible token that has a fixed list of addresses that collectively own the token.
```

## Specification
```solidity
interface IERC1313 {
	// fired anytime a new instance of this token is minted
	// this event **MUST NOT** be fired twice for the same `tokenId`
	event Mint(uint256 indexed tokenId, address indexed nftAddress, uint256 indexed nftTokenId);
	
  // returns the NFT token that owns this token.
	// this function **MUST** throw if the token hasn't been minted yet
	// this function **MUST** always return the same result every time it is called after it has been minted
	// this function **MUST** return the same value as found in the original `Mint` event for the token
	function ownerOf(uint256 index) external view returns (address nftAddress, uint256 nftTokenId);
  
  // returns a current number of mates.
  // this function **MUST** return the latest number of tokens.

>   function current(Counter storage counter) internal view returns (uint256)

Reviewer 2: ```suggestion
  function current(uint256 counter) internal view returns (uint256)
```
Counter must be defined if it is to be used. `uint256` or other primitives are preferred.

  
  // increment number of mates.

>   function increment(Counter storage counter) internal

Reviewer 2: ```suggestion
  function increment(uint256 counter) internal
```

	
	// returns a censorship resistant URI with details about this token instance
	// the tokenUri **MUST** be immutable and content addressable (e.g., ipfs://)
	// the tokenUri **MUST NOT** point at mutable/censorable content (e.g., https://)
	// data from this takes precedence over data returned by `collectionUri`
	// any external links referenced by the content at `tokenUri` also **MUST** follow all of the above rules
	function tokenUri(uint256 tokenId) external view returns (string tokenUri);
  
  // increment number of mates.
  // this function **MUST** 
  function _burn(uint256 tokenId) internal virtual

> }

Reviewer: This code has mixed tabs and spaces, please pick one style and stick with it.

> ```solidity
> interface IERC1313 {
> 	// fired anytime a new instance of this token is minted
> 	// this event **MUST NOT** be fired twice for the same `tokenId`
> 	event Mint(uint256 indexed tokenId, address indexed nftAddress, uint256 indexed nftTokenId);
> 	
>   // returns the NFT token that owns this token.
> 	// this function **MUST** throw if the token hasn't been minted yet
> 	// this function **MUST** always return the same result every time it is called after it has been minted
> 	// this function **MUST** return the same value as found in the original `Mint` event for the token
> 	function ownerOf(uint256 index) external view returns (address nftAddress, uint256 nftTokenId);
>   
>   // returns a current number of mates.
>   // this function **MUST** return the latest number of tokens.
>   function current(Counter storage counter) internal view returns (uint256)
>   
>   // increment number of mates.
>   function increment(Counter storage counter) internal
> 	
> 	// returns a censorship resistant URI with details about this token instance
> 	// the tokenUri **MUST** be immutable and content addressable (e.g., ipfs://)
> 	// the tokenUri **MUST NOT** point at mutable/censorable content (e.g., https://)
> 	// data from this takes precedence over data returned by `collectionUri`
> 	// any external links referenced by the content at `tokenUri` also **MUST** follow all of the above rules
> 	function tokenUri(uint256 tokenId) external view returns (string tokenUri);
>   
>   // increment number of mates.
>   // this function **MUST** 
>   function _burn(uint256 tokenId) internal virtual
> }
> ```

Reviewer: If this is meant to extend EIP-5114, then I recommend something like:
```suggestion

interface IERC1313 is IERC5114 {
  // returns a current number of mates.
  // this function **MUST** return the latest number of tokens.
  function current(Counter storage counter) internal view returns (uint256)
  
  // increment number of mates.
  function increment(Counter storage counter) internal

  // increment number of mates.
  // this function **MUST** 
  function _burn(uint256 tokenId) internal virtual
}
```

## Rationale

### Immutability
To be a countable soul bound token, representing a mate holding a similar SOUL. Intended to be issued without an upper limit.

### Content Addressable URIs Required
Soulbound tokens are meant to be permanent badges/indicators attached to a persona.
This means that not only can the user not transfer ownership, but the minter also cannot withdraw/transfer/change ownership as well.
This includes mutating or removing any remote content as a means of censoring or manipulating specific users.

> 

Reviewer 2: ```suggestion
## Backwards Compatibility

No backward compatibility issues were found.
```
